vault/login_mfa.go

```diff
@@ -1764,25 +1877,53 @@ func formatUsername(format string, alias *identity.Alias, entity *identity.Entit
 	return username
 }
 
-func (c *Core) validateDuo(ctx context.Context, creds []string, mConfig *mfa.Config, username, reqConnectionRemoteAddr string) error {
-	duoConfig := mConfig.GetDuoConfig()
-	if duoConfig == nil {
-		return fmt.Errorf("failed to get Duo configuration for method %q", mConfig.Name)
-	}
+type MFAFactor struct {
+	passcode string
+}
+
+func parseMfaFactors(creds []string) (*MFAFactor, error) {
+	mfaFactor := &MFAFactor{}
+	var err *multierror.Error
 
-	passcode := ""
 	for _, cred := range creds {
-		if strings.HasPrefix(cred, "passcode") {
+		switch {
+		case cred == "": // for the case of push notification
+			continue
+		case strings.HasPrefix(cred, "passcode="):
 			splits := strings.SplitN(cred, "=", 2)
-			if len(splits) != 2 {
-				return fmt.Errorf("invalid credential %q", cred)
-			}
-			if splits[0] == "passcode" {
-				passcode = splits[1]
-			}
+			mfaFactor.passcode = splits[1]
+		case strings.Contains(cred, "="):
+			err = multierror.Append(err, fmt.Errorf("found an invalid MFA cred: %v", cred))
+		default:
+			// a non-empty cred that does not match the above
+			// means it is a passcode
+			mfaFactor.passcode = cred
+		}
+
+		// currently, we only support passcode. Although we probably gain
+		// very little improvement in performance, it still makes sense
+		// to break the for loop
+		if mfaFactor.passcode != "" {
+			break
 		}
 	}
 
+	// don't return an error unless all entries were invalid
+	if mfaFactor.passcode != "" {
```

Comment:
Your current approach is forgiving: so long as we can find something in creds that looks right, you'll use it.  That's an acceptable approach, and maybe the right one.  The downside of being less strict is that it may result in confusing user behaviour,  e.g. garbage was provided and ignored in the past, now I'm putting that garbage in a different position in the creds list and it's failing.

Not asking you to change this unless you agree that a stricter approach might be better.  One reason we might not want to: maybe this isn't the right place.  After all, why do we even have a slice of creds here?  Simply because original the creds came from a header (and still can, in single-phase), and in HTTP headers can be provided multiple times.  There's no legitimate use case that I know of for having more than one creds value for a given mfa method.

Reply:
I think being more strict ends up with better UX, at least users are not confused about it as you described. I will return an error immediately in the second case.
About the creds though, my motivation of using the slice was that if in the future we would want to add other factors, then those would all be in the same creds slice. So, it made sense to loop through creds. 
Having said the above, we could still expect to have one passcode in the creds, and we can be strict about it.